# Тестовое Задание на позицию Infrastructure Data Analyst

## Гайдлайн

Этот документ состоит из 10 заданий, проверяющиx владение инструментами для анализа данных.

ℹ️  Используйте SQL для выполнения `Часть 1` и Python для выполнения `Часть 2`

## Подготовка

Before starting the Test Assignment, please **update the sqlite3** library to the latest version with a code below:

In [1]:
import sqlite3

In [2]:
sqlite3.sqlite_version

'3.36.0'

In [3]:
import pandas as pd
import sqlite3
current_version = sqlite3.sqlite_version
min_version = '3.25.0'

if current_version < min_version:
    msg = f"current version {current_version} is less than {min_version}, update and restart collab runtime"
    raise ImportError(msg)
else:
    msg = f"current version {current_version} is OK!"
    print(msg)

current version 3.36.0 is OK!


## Часть 1: SQL

Используйте только SQL запросы для выполнения этой части.

In [4]:
con = sqlite3.connect('part1_ozon_test_new.db')

df_cities = pd.read_sql_query("SELECT * FROM cities", con)
df_cities

city_name                               city_id
0            Moscow  b00ce276-6dcb-40ee-9016-4109c8c7b60f
1  Saint-Petersburg  ffd147fa-150a-4789-8252-39a4eaf3ee12
2         Krasnodar  737ea994-2e23-4f39-b220-5b9216b56147
3             Adler  a2b8c09f-0ede-491b-98ae-b6dbba89abb8
4      Ekaterinburg  642e49e5-afa1-40af-97ae-fd108818def9

### Q1:
`time_to_delivery` - это метрика, которая считает сколько проходит между созданием (`created_at`) и получением заказа (`delivered_at`)

Посчитайте для каждого города средний `time_to_delivery` и отсорируйте их - от худшего к лучшему

Выведу таблицу **orders** для её изучения, первые 10 строк

In [5]:
df_orders = pd.read_sql_query("SELECT * FROM orders LIMIT 10", con)
df_orders

created_at         delivered_at  value_rub  \
0  2020-02-09 22:21:16  2020-02-11 10:55:42       2753   
1  2020-02-03 04:37:47  2020-02-04 11:30:34       1321   
2  2020-02-09 16:18:59  2020-02-11 00:27:52       1164   
3  2020-02-02 22:32:43  2020-02-03 22:01:54        691   
4  2020-02-08 16:52:01  2020-02-09 21:20:34        727   
5  2020-02-06 01:35:40  2020-02-07 05:36:34       1511   
6  2020-02-01 23:11:21  2020-02-02 23:05:05       3411   
7  2020-02-06 05:18:23  2020-02-07 09:04:41       1605   
8  2020-02-03 23:04:46  2020-02-04 22:43:19       2092   
9  2020-02-05 06:13:53  2020-02-06 16:51:50       1314   

                                user_id                               city_id  
0  ec3c583d-2098-4c13-8456-72980e9943b7  b00ce276-6dcb-40ee-9016-4109c8c7b60f  
1  a0bb6ea3-4128-40ed-b0ee-c75f82dcbe6f  b00ce276-6dcb-40ee-9016-4109c8c7b60f  
2  1f9cf58e-6751-4ad8-854e-e3e50e00b6b6  b00ce276-6dcb-40ee-9016-4109c8c7b60f  
3  90dc4973-50e3-48e1-9655-7cbaa496aac7  b00ce276-6dcb-40ee-9016-4109c8c7b60f  
4  82fcbbde-8a22-46fb-bd2b-376781b3a417  b00ce276-6dcb-40ee-9016-4109c8c7b60f  
5  52faa6ae-3cdb-48c3-8a10-3e330039d16e  b00ce276-6dcb-40ee-9016-4109c8c7b60f  
6  4746b9ef-b0e7-45fb-945a-177aba3e06e0  b00ce276-6dcb-40ee-9016-4109c8c7b60f  
7  e79147ad-8c4f-44f7-8659-399e0942e456  b00ce276-6dcb-40ee-9016-4109c8c7b60f  
8  20a9848e-c1b8-4e71-af42-7b1bd0627926  b00ce276-6dcb-40ee-9016-4109c8c7b60f  
9  3f91ad1b-b1cc-4929-8f6f-a9da86176892  b00ce276-6dcb-40ee-9016-4109c8c7b60f

Нахожу разницу во вререни между созданием **created_at** и получением заказа **delivered_at** в часах и отсортирую их в порядке убывания

In [6]:
df_time_to_delivery = pd.read_sql_query("SELECT user_id, created_at, delivered_at, ROUND((julianday(delivered_at) - julianday(created_at))*24, 2) AS time_to_delivery_user_id FROM orders ORDER BY time_to_delivery_user_id DESC LIMIT 10", con)
df_time_to_delivery

user_id                  created_at  \
0  063864ea-d96a-4f6b-ada1-b6dbcef9e70d         2020-02-07 21:05:27   
1  29abcfbb-e9dc-4e89-822c-43764ad0db5e         2020-02-09 21:10:55   
2  9697d0b1-2e5b-40b4-aac0-d5df85c936f3  2020-02-18 10:16:37.339581   
3  eeda8ee5-402f-44b6-9c12-e10af4d4689e  2020-02-10 14:48:32.097453   
4  c3466edf-2230-4f6c-95cd-94201ee39557         2020-02-09 23:31:02   
5  41082976-c784-41ee-82b4-2b21898e8d57         2020-02-04 06:09:15   
6  5513fc2a-f89f-4c0e-81d2-b9f880d83e06         2020-02-08 00:16:22   
7  9697d0b1-2e5b-40b4-aac0-d5df85c936f3         2020-02-01 06:40:33   
8  a1b834ad-6068-4038-a605-05032126bd7c         2020-02-06 16:44:23   
9  65a3b6c7-54b6-4ea5-ac45-6410052cefbe  2020-02-10 00:56:20.716155   

          delivered_at  time_to_delivery_user_id  
0  2020-02-11 20:23:21                     95.30  
1  2020-02-13 19:35:35                     94.41  
2  2020-02-22 05:53:44                     91.62  
3  2020-02-14 09:09:12                     90.34  
4  2020-02-13 16:19:36                     88.81  
5  2020-02-07 18:18:37                     84.16  
6  2020-02-11 11:11:32                     82.92  
7  2020-02-04 14:48:40                     80.14  
8  2020-02-10 00:19:09                     79.58  
9  2020-02-13 08:01:00                     79.08

теперь группирую данный результат по городам и с помощью агрегирующей функции AVG() нахожу среденее время ожидания заказа

In [7]:
df_time_to_delivery_city = pd.read_sql_query("SELECT city_id, AVG(ROUND((julianday(delivered_at) - julianday(created_at))*24, 2)) AS time_to_delivery FROM orders GROUP BY city_id ORDER BY time_to_delivery DESC", con)
df_time_to_delivery_city

city_id  time_to_delivery
0  a2b8c09f-0ede-491b-98ae-b6dbba89abb8         73.677727
1  642e49e5-afa1-40af-97ae-fd108818def9         62.270845
2  ffd147fa-150a-4789-8252-39a4eaf3ee12         39.539220
3  b00ce276-6dcb-40ee-9016-4109c8c7b60f         30.005725

**city_id** не показательное поле, поэтому объединим таблицы по нему и выведем название городов

In [8]:
df_time_to_delivery_city = pd.read_sql_query("SELECT cities.city_name, AVG(ROUND((julianday(delivered_at) - julianday(created_at))*24, 2)) AS time_to_delivery FROM orders JOIN cities ON orders.city_id = cities.city_id GROUP BY city_name ORDER BY time_to_delivery DESC", con)
df_time_to_delivery_city

city_name  time_to_delivery
0             Adler         73.677727
1      Ekaterinburg         62.270845
2  Saint-Petersburg         39.539220
3            Moscow         30.005725

Можно заметить что в среднем в городе в Москве заказы доставляют в два раза быстрее чем в Екатеринбурге и почти в 2.5 раза быстрее, чем в Адлере, это можно связать сболее развитой инфраструктурой города, и скорее всего колличество заказов в Москве куда больше, а следовательно среднее значение может быть непоказательным в данном случае.

Также можно увидеть, что в таблице отсутствует Краснодар, изучу данные получше и узнаем сколько всего заказов было оформлено в Краснодаре

In [9]:
df_city_count = pd.read_sql_query("SELECT cities.city_name, COUNT(*) FROM orders JOIN cities ON orders.city_id = cities.city_id GROUP BY cities.city_name", con)
df_city_count

city_name  COUNT(*)
0             Adler        22
1      Ekaterinburg        71
2            Moscow       269
3  Saint-Petersburg       141

In [10]:
df_city_count_sum = pd.read_sql_query("SELECT COUNT(*) FROM orders", con)
df_city_count_sum

COUNT(*)
0       503

In [11]:
df_city_count_sum = pd.read_sql_query("SELECT SUM(table_1.cnt) FROM (SELECT cities.city_name, COUNT(*) as cnt FROM orders JOIN cities ON orders.city_id = cities.city_id GROUP BY cities.city_name) AS table_1", con)
df_city_count_sum

SUM(table_1.cnt)
0               503

Действительно в базе данных нет ни одного заказа из Краснодара, возможно это ошибка, раз город попал в базу, либо по каким-то причинам в Краснодаре в этот период оформление заказов было недоступно. 

In [12]:
df_time_to_delivery_city

city_name  time_to_delivery
0             Adler         73.677727
1      Ekaterinburg         62.270845
2  Saint-Petersburg         39.539220
3            Moscow         30.005725

### Q2: 
А сколько в среднем проходит между тем как пользователь получил первый заказ, и сделал следующий заказ?

Для решения этой задачи предлагаю определить самую раннюю и последнюю дату, то есть для каждого пользователя определю MIN и MAX даты и узнаем сколько каждый пользователь оофрмлял заказов за этот период, но перед нахождением среднего надо вычесть из разности ранней и поздней дат время ожидания заказов, но мы уже имеем несколько заготовок в виде таблиц **df_time_to_delivery** отсортированной по каждому пользователю

In [13]:
## то есть смысл нахождения AVG() такой, допустим пользователь оформил два заказа
## первый заказ в 12:10 и получил его в 12:40
## второй заказ в 15:45 и получил его в 16:00

## Мы прекрасно понимаем что по условию задачи нам нужно найти время (15:45 - 12:40) и получить ответ: 3 часа 5 минут

## но в sql это сделать можно следующим образом: находим разницу между ранней и поздней датами => (16:00 - 12:10) = 3 часа 50 минут

## теперь из этой разницы вычтем сумму всех ожиданий клиента => 3часа 50 минут  - ((12:40 - 12:10 = 30 минут) + 
## + (16:00 - 15:45 = 15 минут) = 45 минут) = 3 часа 5 минут

## Значит задача решена верно, применим это уже с реальными данными

In [14]:
df_user_id = pd.read_sql_query("SELECT user_id, ROUND((julianday(delivered_at) - julianday(created_at))*24, 2) AS time_to_delivery_user_id FROM orders ORDER BY user_id DESC LIMIT 10", con)
df_user_id

user_id  time_to_delivery_user_id
0  fe6baf5c-c617-4e0b-9ec0-b0c350b2ff45                     33.13
1  fe6baf5c-c617-4e0b-9ec0-b0c350b2ff45                     34.93
2  fb9bf9ec-b526-4d8c-91ba-f1bfcea16a85                     37.29
3  fb35e63b-da01-468f-bc79-596b788aebfb                     19.51
4  fb35e63b-da01-468f-bc79-596b788aebfb                     27.71
5  fb35e63b-da01-468f-bc79-596b788aebfb                     31.18
6  fb35e63b-da01-468f-bc79-596b788aebfb                     30.88
7  fa9c535e-7abd-4e41-812d-236274010313                     47.39
8  f865b031-b3f5-47ad-aabf-96651cec8d4c                     30.28
9  f73a7b7a-a961-49ca-b23e-c8191306ef1b                     26.34

In [15]:
df_user_id = pd.read_sql_query("SELECT user_id, MIN(created_at), MAX(delivered_at), ROUND((julianday(MAX(delivered_at))-julianday(MIN(created_at)))*24,2) AS diff_max_man,COUNT(user_id) AS cnt, sum(ROUND((julianday(delivered_at) - julianday(created_at))*24, 2)) AS sum_time_to_delivery_user_id FROM orders GROUP BY user_id ORDER BY user_id DESC LIMIT 10", con)
df_user_id

user_id             MIN(created_at)  \
0  fe6baf5c-c617-4e0b-9ec0-b0c350b2ff45         2020-02-03 23:53:43   
1  fb9bf9ec-b526-4d8c-91ba-f1bfcea16a85         2020-02-09 05:55:27   
2  fb35e63b-da01-468f-bc79-596b788aebfb         2020-02-09 04:28:27   
3  fa9c535e-7abd-4e41-812d-236274010313         2020-02-03 20:46:00   
4  f865b031-b3f5-47ad-aabf-96651cec8d4c         2020-02-07 14:13:15   
5  f73a7b7a-a961-49ca-b23e-c8191306ef1b         2020-02-03 05:02:33   
6  f64d3ec4-0f61-4f00-a9f4-cd1149e645ba         2020-02-02 10:40:47   
7  f5c71a71-b525-4406-bdbc-49b5e7cb3710         2020-02-03 01:00:58   
8  f45deeb4-391b-49ad-a777-bc8120931983  2020-02-08 18:30:00.478189   
9  f42d4de5-2fa0-4c06-be16-5b158e1e71f6         2020-02-03 17:45:59   

     MAX(delivered_at)  diff_max_man  cnt  sum_time_to_delivery_user_id  
0  2020-02-07 22:43:36         94.83    2                         68.06  
1  2020-02-10 19:12:40         37.29    1                         37.29  
2  2020-02-15 03:41:11        143.21    4                        109.28  
3  2020-02-05 20:09:13         47.39    1                         47.39  
4  2020-02-08 20:30:07         30.28    1                         30.28  
5  2020-02-10 20:07:09        183.08    4                        113.33  
6  2020-02-08 14:35:00        147.90    4                        107.45  
7  2020-02-03 22:06:55         21.10    1                         21.10  
8  2020-02-14 14:45:57        140.27    3                         71.70  
9  2020-02-13 02:49:09        225.05    4                        184.06

Данные получены, теперь нужно лишь найти разницу столбцов и поделить на колличество заказов

Можно заметить, что пользователей с одним заказом нужно удалить из данных, так как по сути у него нет той метрики, которую просят найти в задаче, но их данные в столбцах **diff_max_man** и **sum_time_to_delivery_user_id** равны и в итоге они получат 0 в ячейке, что будет означать отсутствие других закозов после первого

In [16]:
df_user_id = pd.read_sql_query("SELECT user_id,((diff_max_man-sum_time_to_delivery_user_id)/cnt) AS avg_time FROM(SELECT user_id, MIN(created_at), MAX(delivered_at), ROUND((julianday(MAX(delivered_at))-julianday(MIN(created_at)))*24,2) AS diff_max_man,COUNT(user_id) AS cnt, sum(ROUND((julianday(delivered_at) - julianday(created_at))*24, 2)) AS sum_time_to_delivery_user_id FROM orders GROUP BY user_id ORDER BY user_id DESC) LIMIT 10", con)
df_user_id

user_id   avg_time
0  fe6baf5c-c617-4e0b-9ec0-b0c350b2ff45  13.385000
1  fb9bf9ec-b526-4d8c-91ba-f1bfcea16a85   0.000000
2  fb35e63b-da01-468f-bc79-596b788aebfb   8.482500
3  fa9c535e-7abd-4e41-812d-236274010313   0.000000
4  f865b031-b3f5-47ad-aabf-96651cec8d4c   0.000000
5  f73a7b7a-a961-49ca-b23e-c8191306ef1b  17.437500
6  f64d3ec4-0f61-4f00-a9f4-cd1149e645ba  10.112500
7  f5c71a71-b525-4406-bdbc-49b5e7cb3710   0.000000
8  f45deeb4-391b-49ad-a777-bc8120931983  22.856667
9  f42d4de5-2fa0-4c06-be16-5b158e1e71f6  10.247500

Для наглядности взглянем на эту таблицу без нулей

In [17]:
df_user_id = pd.read_sql_query("SELECT user_id,((diff_max_man-sum_time_to_delivery_user_id)/cnt) AS avg_time FROM(SELECT user_id, MIN(created_at), MAX(delivered_at), ROUND((julianday(MAX(delivered_at))-julianday(MIN(created_at)))*24,2) AS diff_max_man,COUNT(user_id) AS cnt, sum(ROUND((julianday(delivered_at) - julianday(created_at))*24, 2)) AS sum_time_to_delivery_user_id FROM orders GROUP BY user_id ORDER BY user_id DESC) WHERE diff_max_man != sum_time_to_delivery_user_id LIMIT 10", con)
df_user_id

user_id   avg_time
0  fe6baf5c-c617-4e0b-9ec0-b0c350b2ff45  13.385000
1  fb35e63b-da01-468f-bc79-596b788aebfb   8.482500
2  f73a7b7a-a961-49ca-b23e-c8191306ef1b  17.437500
3  f64d3ec4-0f61-4f00-a9f4-cd1149e645ba  10.112500
4  f45deeb4-391b-49ad-a777-bc8120931983  22.856667
5  f42d4de5-2fa0-4c06-be16-5b158e1e71f6  10.247500
6  f2b762f8-e621-4173-bbe7-c517e2659f70  17.403333
7  f29ae1ca-85af-4422-9515-33e8a40b980b   1.855000
8  efb0359e-5535-4e34-8047-ea29be5e420c   9.180000
9  ef532da2-9479-42b8-9023-68ef3dc0249a   4.500000

### Q3: 

Посчитайте сколько всего пользователей в Москве имеют средний чек выше 2000 рублей (в % от всех пользователей, у которых был хотя бы один заказ в Москве)

In [18]:
df_id_moscow = pd.read_sql_query("SELECT city_id FROM cities WHERE city_name LIKE '%Moscow%'", con)
df_id_moscow

city_id
0  b00ce276-6dcb-40ee-9016-4109c8c7b60f

In [19]:
df_user_id_msc = pd.read_sql_query("SELECT * FROM orders WHERE city_id IN (SELECT city_id FROM cities WHERE city_name LIKE '%Moscow%') LIMIT 10", con)
df_user_id_msc

created_at         delivered_at  value_rub  \
0  2020-02-09 22:21:16  2020-02-11 10:55:42       2753   
1  2020-02-03 04:37:47  2020-02-04 11:30:34       1321   
2  2020-02-09 16:18:59  2020-02-11 00:27:52       1164   
3  2020-02-02 22:32:43  2020-02-03 22:01:54        691   
4  2020-02-08 16:52:01  2020-02-09 21:20:34        727   
5  2020-02-06 01:35:40  2020-02-07 05:36:34       1511   
6  2020-02-01 23:11:21  2020-02-02 23:05:05       3411   
7  2020-02-06 05:18:23  2020-02-07 09:04:41       1605   
8  2020-02-03 23:04:46  2020-02-04 22:43:19       2092   
9  2020-02-05 06:13:53  2020-02-06 16:51:50       1314   

                                user_id                               city_id  
0  ec3c583d-2098-4c13-8456-72980e9943b7  b00ce276-6dcb-40ee-9016-4109c8c7b60f  
1  a0bb6ea3-4128-40ed-b0ee-c75f82dcbe6f  b00ce276-6dcb-40ee-9016-4109c8c7b60f  
2  1f9cf58e-6751-4ad8-854e-e3e50e00b6b6  b00ce276-6dcb-40ee-9016-4109c8c7b60f  
3  90dc4973-50e3-48e1-9655-7cbaa496aac7  b00ce276-6dcb-40ee-9016-4109c8c7b60f  
4  82fcbbde-8a22-46fb-bd2b-376781b3a417  b00ce276-6dcb-40ee-9016-4109c8c7b60f  
5  52faa6ae-3cdb-48c3-8a10-3e330039d16e  b00ce276-6dcb-40ee-9016-4109c8c7b60f  
6  4746b9ef-b0e7-45fb-945a-177aba3e06e0  b00ce276-6dcb-40ee-9016-4109c8c7b60f  
7  e79147ad-8c4f-44f7-8659-399e0942e456  b00ce276-6dcb-40ee-9016-4109c8c7b60f  
8  20a9848e-c1b8-4e71-af42-7b1bd0627926  b00ce276-6dcb-40ee-9016-4109c8c7b60f  
9  3f91ad1b-b1cc-4929-8f6f-a9da86176892  b00ce276-6dcb-40ee-9016-4109c8c7b60f

In [20]:
df_user_id_msc = pd.read_sql_query("SELECT COUNT(*) FROM orders WHERE city_id IN (SELECT city_id FROM cities WHERE city_name LIKE '%Moscow%')", con)
df_user_id_msc

COUNT(*)
0       269

По условию нас интересует средний чек, тогда отсортируем заказы по Москве и найдём колличество заказов и полную сумму заказов таких пользователей

In [21]:
df_user_id_total = pd.read_sql_query("SELECT user_id, SUM(value_rub) AS total, COUNT(user_id) AS cnt_id, (SUM(value_rub)/COUNT(user_id)) AS avg_total FROM orders WHERE city_id IN (SELECT city_id FROM cities WHERE city_name LIKE '%Moscow%') GROUP BY user_id LIMIT 10", con)
df_user_id_total

user_id  total  cnt_id  avg_total
0  01df03f8-62d7-4617-82a3-5efa32c88887   2685       1       2685
1  03af905e-4600-4414-90d4-f3cbdd42e369   2229       1       2229
2  0408019e-065f-47a8-9680-6ebcbef49eb1   3225       1       3225
3  0513f3bc-7cd1-408c-9b3e-d456ec0122bd   1621       1       1621
4  05dcd22c-005f-4bcf-ba8c-332cc2d6233e   1794       1       1794
5  08aef604-91a3-4cb5-8adf-24aab4c2322d   1240       1       1240
6  090555ee-5d40-4b04-84dd-f552ee4076e5   2429       2       1214
7  0ac95090-362e-4bf7-8c5d-d6e9c36a88c4   1852       2        926
8  0c741992-45fc-4e6b-9ce8-b57c686dae4b   3407       2       1703
9  0d8a152d-2bcd-48ff-9683-207824333219   7916       3       2638

Отлично, теперь из этой таблицы нужны две выборки:

1) количесвто всех клиентов по id

2) количесвто клиентов чьё значение total > 2000

In [22]:
df_user_cnt_id_msc = pd.read_sql_query("SELECT COUNT(*) AS cnt_id_msc FROM(SELECT user_id, SUM(value_rub) AS total, COUNT(user_id) AS cnt_id, (SUM(value_rub)/COUNT(user_id)) AS avg_total FROM orders WHERE city_id IN (SELECT city_id FROM cities WHERE city_name LIKE '%Moscow%') GROUP BY user_id)", con)
df_user_cnt_id_msc

cnt_id_msc
0         150

In [23]:
df_user_id_total = pd.read_sql_query("SELECT COUNT(user_id) AS cnt_msc_more_2000 FROM (SELECT user_id, SUM(value_rub) AS total, COUNT(user_id) AS cnt_id, (SUM(value_rub)/COUNT(user_id)) AS avg_total FROM orders WHERE city_id IN (SELECT city_id FROM cities WHERE city_name LIKE '%Moscow%') GROUP BY user_id) WHERE avg_total > 2000 ", con)
df_user_id_total

cnt_msc_more_2000
0                 59

In [24]:
df_user_cnt_msc_more_2000 = pd.read_sql_query("SELECT (100*(SELECT COUNT(user_id) AS cnt_msc_more_2000 FROM (SELECT user_id, SUM(value_rub) AS total, COUNT(user_id) AS cnt_id, (SUM(value_rub)/COUNT(user_id)) AS avg_total FROM orders WHERE city_id IN (SELECT city_id FROM cities WHERE city_name LIKE '%Moscow%') GROUP BY user_id) WHERE avg_total > 2000 )/(SELECT COUNT(*) AS cnt_id_msc FROM(SELECT user_id, SUM(value_rub) AS total, COUNT(user_id) AS cnt_id, (SUM(value_rub)/COUNT(user_id)) AS avg_total FROM orders WHERE city_id IN (SELECT city_id FROM cities WHERE city_name LIKE '%Moscow%') GROUP BY user_id))) AS perc_of_users", con)
df_user_cnt_msc_more_2000

perc_of_users
0             39

Итого 39% заказчиков из Москвы в среднем тратят больше 2000 рублей за один лишь заказ

### Q4:
Посчитайте сколько часов в среднем проходит между заказами одного и того же пользователя?

Это задача аналогичная второй задаче, только будем использовать только время заказов без времени получения этих самых заказов

Поэтому используем старую таблицу о заказах пользователей с учётом замечаний

In [25]:
df_user_id = pd.read_sql_query("SELECT user_id, MIN(created_at), MAX(created_at), ROUND((julianday(MAX(created_at))-julianday(MIN(created_at)))*24,2) AS diff_max_man,COUNT(user_id) AS cnt FROM orders GROUP BY user_id ORDER BY user_id DESC LIMIT 10", con)
df_user_id

user_id             MIN(created_at)  \
0  fe6baf5c-c617-4e0b-9ec0-b0c350b2ff45         2020-02-03 23:53:43   
1  fb9bf9ec-b526-4d8c-91ba-f1bfcea16a85         2020-02-09 05:55:27   
2  fb35e63b-da01-468f-bc79-596b788aebfb         2020-02-09 04:28:27   
3  fa9c535e-7abd-4e41-812d-236274010313         2020-02-03 20:46:00   
4  f865b031-b3f5-47ad-aabf-96651cec8d4c         2020-02-07 14:13:15   
5  f73a7b7a-a961-49ca-b23e-c8191306ef1b         2020-02-03 05:02:33   
6  f64d3ec4-0f61-4f00-a9f4-cd1149e645ba         2020-02-02 10:40:47   
7  f5c71a71-b525-4406-bdbc-49b5e7cb3710         2020-02-03 01:00:58   
8  f45deeb4-391b-49ad-a777-bc8120931983  2020-02-08 18:30:00.478189   
9  f42d4de5-2fa0-4c06-be16-5b158e1e71f6         2020-02-03 17:45:59   

              MAX(created_at)  diff_max_man  cnt  
0  2020-02-06 11:47:32.955526         59.90    2  
1         2020-02-09 05:55:27          0.00    1  
2  2020-02-13 20:48:21.241216        112.33    4  
3         2020-02-03 20:46:00          0.00    1  
4         2020-02-07 14:13:15          0.00    1  
5  2020-02-09 10:59:07.356538        149.94    4  
6  2020-02-07 16:11:27.196179        125.51    4  
7         2020-02-03 01:00:58          0.00    1  
8  2020-02-13 16:32:37.933901        118.04    3  
9  2020-02-11 06:30:55.035819        180.75    4

Сразу же избавимся от нулей в данных, так как пользователи оформившие один заказ будет непоказательны в этой статистике

In [26]:
df_user_id = pd.read_sql_query("SELECT ROUND((diff_max_man/cnt),2) AS avg_time_created_at FROM(SELECT user_id, MIN(created_at), MAX(created_at), ROUND((julianday(MAX(created_at))-julianday(MIN(created_at)))*24,2) AS diff_max_man,COUNT(user_id) AS cnt FROM orders GROUP BY user_id ORDER BY user_id DESC) WHERE diff_max_man!=0 LIMIT 10", con)
df_user_id

avg_time_created_at
0                29.95
1                28.08
2                37.49
3                31.38
4                39.35
5                45.19
6                54.32
7                26.61
8                45.03
9                19.82

Отсортирую данные по возрастанию и добавлю **user_id**

In [27]:
df_user_id = pd.read_sql_query("SELECT user_id, ROUND((diff_max_man/cnt),2) AS avg_time_created_at FROM(SELECT user_id, MIN(created_at), MAX(created_at), ROUND((julianday(MAX(created_at))-julianday(MIN(created_at)))*24,2) AS diff_max_man,COUNT(user_id) AS cnt FROM orders GROUP BY user_id ORDER BY user_id DESC) WHERE diff_max_man!=0 ORDER BY avg_time_created_at LIMIT 10", con)
df_user_id

user_id  avg_time_created_at
0  466c14f4-3091-4bdf-9553-76ea022a7788                 4.99
1  0c741992-45fc-4e6b-9ce8-b57c686dae4b                11.36
2  b931b8c6-1bc2-4df7-bc97-da95aade9fe8                12.37
3  16057d77-9edd-47cb-91e5-4bcd83d89edb                13.33
4  83b64228-8704-4d28-b676-1dd79b037e4e                13.69
5  ba9bd33c-3c3c-4307-b173-7224c927a0f6                13.74
6  cd7bb72b-e1fe-42f8-81f6-3e5e6a5fd254                14.04
7  61e5a742-640c-4b13-8e05-38b5175e6a58                15.49
8  d12c7b7c-2f1c-4afc-9224-5bb0843cdd4c                16.28
9  996aa60a-a4d2-4bce-aefb-3d40acd1c695                16.79

Ого, пользователь *466c14f4-3091-4bdf-9553-76ea022a7788* в среднем делает один заказ в 5 часов

Конечно было бы лучше узнать, что у него больше 2ух заказов, но если он ещё и попал в группу пользователей с чеком от 2000, то такой клиент очень ценный, но это уже другая задача.


## Часть 2: requests & pandas

Вторая часть проверяет владение `requests`, `pandas` и навыки визуализации

* для q5, q6 исползуйте модуль `requests`
* для q8, q9 используйте `pandas`
* для q7 используйте любую библиотеку визаулизации на ваш выбор. Я советую `altair` или `plotly`

Перед выполнением задания пройдите requests quickstart
*   http://bit.ly/2KoI43m

Мы будем пользоваться общедоступным REST API для тестирования: https://randomuser.me/api/

Это API endpoint, который возвращает случайно сгенерированно пользователя.


In [28]:
import requests

### Q5 

Скачайте и красиво распечатайте результат для случайного пользователя 

(подсказка: пользуйтесь GET методом)

In [29]:
URL = 'https://randomuser.me/api/'
req = requests.get(URL)
print(req.text)

{"results":[{"gender":"male","name":{"title":"Mr","first":"Pablo","last":"Vargas"},"location":{"street":{"number":261,"name":"Calle de Bravo Murillo"},"city":"Santander","state":"Andalucía","country":"Spain","postcode":32374,"coordinates":{"latitude":"-31.4244","longitude":"-12.6297"},"timezone":{"offset":"+5:30","description":"Bombay, Calcutta, Madras, New Delhi"}},"email":"pablo.vargas@example.com","login":{"uuid":"dfb06c03-765c-4ad9-a0c0-8720eaf416a1","username":"smallbear595","password":"7734","salt":"ExwYM3JO","md5":"80bcec5141a17b0c218e35cf66472fd8","sha1":"1a711fe6709280e70a5e10a5b9f8e3746a3f0f0c","sha256":"0f8e9e1f031714b63c47a49bb76957b9dacc42df898c51a29ec674e0bf51ebc5"},"dob":{"date":"1972-03-06T04:37:15.040Z","age":50},"registered":{"date":"2004-11-16T13:30:03.756Z","age":17},"phone":"996-920-460","cell":"697-217-787","id":{"name":"DNI","value":"79349394-C"},"picture":{"large":"https://randomuser.me/api/portraits/men/5.jpg","medium":"https://randomuser.me/api/portraits/med/m

In [30]:
import json

BASE_URL = 'https://randomuser.me/api/'
response = requests.get(BASE_URL)
response_parsed = json.loads(response.text)
print(response_parsed['info'])

{'seed': '85441cb240b00920', 'results': 1, 'page': 1, 'version': '1.4'}


In [31]:
list_user=req.text

In [32]:
list_user_json = json.loads(list_user)
list_user_json

{'results': [{'gender': 'male',
   'name': {'title': 'Mr', 'first': 'Pablo', 'last': 'Vargas'},
   'location': {'street': {'number': 261, 'name': 'Calle de Bravo Murillo'},
    'city': 'Santander',
    'state': 'Andalucía',
    'country': 'Spain',
    'postcode': 32374,
    'coordinates': {'latitude': '-31.4244', 'longitude': '-12.6297'},
    'timezone': {'offset': '+5:30',
     'description': 'Bombay, Calcutta, Madras, New Delhi'}},
   'email': 'pablo.vargas@example.com',
   'login': {'uuid': 'dfb06c03-765c-4ad9-a0c0-8720eaf416a1',
    'username': 'smallbear595',
    'password': '7734',
    'salt': 'ExwYM3JO',
    'md5': '80bcec5141a17b0c218e35cf66472fd8',
    'sha1': '1a711fe6709280e70a5e10a5b9f8e3746a3f0f0c',
    'sha256': '0f8e9e1f031714b63c47a49bb76957b9dacc42df898c51a29ec674e0bf51ebc5'},
   'dob': {'date': '1972-03-06T04:37:15.040Z', 'age': 50},
   'registered': {'date': '2004-11-16T13:30:03.756Z', 'age': 17},
   'phone': '996-920-460',
   'cell': '697-217-787',
   'id': {'name':

Теперь напишу функцию, для понятной подачи информации из json формата

In [33]:
from PIL import Image

def fync_user(list_user_json):
    list_1 = list_user_json.get('results')
    print('name:', (list_1[0].get('name')).get('title'),(list_1[0].get('name')).get('first'), (list_1[0].get('name')).get('last'))
    print('gender:', list_1[0].get('gender'))
    print('age:',(list_1[0].get('dob')).get('age'))
    print('dob:',(list_1[0].get('dob')).get('date').split('T')[0])
    print('country:', (list_1[0].get('location')).get('country'),
         '\nstate -', (list_1[0].get('location')).get('state'),
         '\ncity -',(list_1[0].get('location')).get('city'),
         '\nstreet -',((list_1[0].get('location')).get('street')).get('name'),
         ((list_1[0].get('location')).get('street')).get('number'),
         '\ncoordinates:','latitude =', ((list_1[0].get('location')).get('coordinates')).get('latitude'),
         'longitude =', ((list_1[0].get('location')).get('coordinates')).get('longitude'),
         '\ntimezone:', ((list_1[0].get('location')).get('timezone')).get('offset'))
    print('email:', list_1[0].get('email'))
    print('login:', '\nusername - ',(list_1[0].get('login')).get('username'),
         '\npassword - ',(list_1[0].get('login')).get('password'),
         '\nuuid - ',(list_1[0].get('login')).get('uuid'))
    print('registered:', 'date -', (list_1[0].get('registered')).get('date').split('T')[0], (list_1[0].get('registered')).get('date').split('T')[1])
    print('phone:', list_1[0].get('phone'),
         '\ncell:', list_1[0].get('cell'))
    print('id:','name -', list_1[0].get('id').get('name'), '\nvalue -', list_1[0].get('id').get('value'))
    url_1= list_1[0].get('picture').get('large')
    res_1= requests.get(url_1, stream =True).raw
    img_1=Image.open(res_1)
    img_1.show()
    print('nat:', list_1[0].get('nat'))

Функция готова, осталось проверить как она работает для случайного пользователя

In [34]:
fync_user(list_user_json)

name: Mr Pablo Vargas
gender: male
age: 50
dob: 1972-03-06
country: Spain 
state - Andalucía 
city - Santander 
street - Calle de Bravo Murillo 261 
coordinates: latitude = -31.4244 longitude = -12.6297 
timezone: +5:30
email: pablo.vargas@example.com
login: 
username -  smallbear595 
password -  7734 
uuid -  dfb06c03-765c-4ad9-a0c0-8720eaf416a1
registered: date - 2004-11-16 13:30:03.756Z
phone: 996-920-460 
cell: 697-217-787
id: name - DNI 
value - 79349394-C
nat: ES


### Q6

Однако, для предсказумых результатов можно передавать в запрос параметр `seed`. Тогда всегда будет возвращаться один и тот же пользователь, например:

* seed `65bfda4df705bed3` всегда возвращает Selma Mаdsen, 61 год
* seed `1f6bcdddf53702b0` всегда будет возвращать Deivid Barros из Бразилии

Мы подготовили для вас 20 сидов, ваша задача:

1. Скачать 20 пользователей с вышеуказанными сидами
2. положить их в pd.DataFrame в удобном виде. Должно получиться примерно так:

![](http://i.imgur.com/cS0Ma69.png)

In [35]:
random_seeds = ['39c74702ac1bc239', '4c704ac2e3bafcd0', 'bb4fb271a935393b',
 'f1aa6e19de4c5f35', 'f62795d3059338ca', 'a72f8ae890ef4b40', '7fd112b2157130ed',
 'cc871caae4b8b48f', '58dcbce92a0a001a', 'b1942240f0ca789a', '3319a0fc434de151',
 '93c3d966089ee076', '1a04f96f90069e18', '8a7a8ed9a429df61', '85675308636fbe7e',
 'e232bf0460d651e3', '937867e6ea5d9edb', '349e4916c919edbb', 'c230b20320c3f59f',
 'fed1ecb439e22534']

Разберёмся с **random_seeds**

In [36]:
for i in random_seeds:
    BASE_URL_new = 'https://randomuser.me/api/?seed='
    BASE_URL_i = BASE_URL_new + i
    response = requests.get(BASE_URL_i)
    response_parsed = json.loads(response.text)
    print(response_parsed['info'])

{'seed': '39c74702ac1bc239', 'results': 1, 'page': 1, 'version': '1.4'}
{'seed': '4c704ac2e3bafcd0', 'results': 1, 'page': 1, 'version': '1.4'}
{'seed': 'bb4fb271a935393b', 'results': 1, 'page': 1, 'version': '1.4'}
{'seed': 'f1aa6e19de4c5f35', 'results': 1, 'page': 1, 'version': '1.4'}
{'seed': 'f62795d3059338ca', 'results': 1, 'page': 1, 'version': '1.4'}
{'seed': 'a72f8ae890ef4b40', 'results': 1, 'page': 1, 'version': '1.4'}
{'seed': '7fd112b2157130ed', 'results': 1, 'page': 1, 'version': '1.4'}
{'seed': 'cc871caae4b8b48f', 'results': 1, 'page': 1, 'version': '1.4'}
{'seed': '58dcbce92a0a001a', 'results': 1, 'page': 1, 'version': '1.4'}
{'seed': 'b1942240f0ca789a', 'results': 1, 'page': 1, 'version': '1.4'}
{'seed': '3319a0fc434de151', 'results': 1, 'page': 1, 'version': '1.4'}
{'seed': '93c3d966089ee076', 'results': 1, 'page': 1, 'version': '1.4'}
{'seed': '1a04f96f90069e18', 'results': 1, 'page': 1, 'version': '1.4'}
{'seed': '8a7a8ed9a429df61', 'results': 1, 'page': 1, 'version':

Пока рано приступать к созданию дата фрейма

создам из этого цикла списки и словари со значениями, которые станут основой для таблицы

In [37]:
dict_data = {}
for i in random_seeds:
    BASE_URL_new = 'https://randomuser.me/api/?seed='
    BASE_URL_i = BASE_URL_new + i
    response = requests.get(BASE_URL_i)
    response_parsed = json.loads(response.text)
    list_i = response_parsed.get('results')
    dict_data[i]=[list_i[0].get('gender'), list_i[0].get('name'),list_i[0].get('location'),
                  list_i[0].get('email'),list_i[0].get('login'),
                 list_i[0].get('dob'), list_i[0].get('registered'),
                 list_i[0].get('phone'), list_i[0].get('cell'),
                 list_i[0].get('id')]
    

In [38]:
dict_data

{'39c74702ac1bc239': ['male',
  {'title': 'Mr', 'first': 'Sean', 'last': 'Weaver'},
  {'street': {'number': 9198, 'name': 'Shady Ln Dr'},
   'city': 'Hampton',
   'state': 'Illinois',
   'country': 'United States',
   'postcode': 79680,
   'coordinates': {'latitude': '82.9441', 'longitude': '-71.6443'},
   'timezone': {'offset': '+4:00',
    'description': 'Abu Dhabi, Muscat, Baku, Tbilisi'}},
  'sean.weaver@example.com',
  {'uuid': 'dfaeb64c-3da9-4034-81ad-c79b109fea54',
   'username': 'redswan825',
   'password': 'austin',
   'salt': 'l1PMcmqF',
   'md5': '111f2eef8111d0bb4133f86c0f12d24e',
   'sha1': '608754836a8c91faa67c11a1c228893ae502fdc9',
   'sha256': 'd7466b98b5df24b4cad9556ab795bf2e6b1bc93ce718a5a6203474dc549cb8df'},
  {'date': '1972-06-19T14:17:59.978Z', 'age': 50},
  {'date': '2014-08-03T07:25:36.795Z', 'age': 8},
  '(206) 996-3299',
  '(851) 950-2650',
  {'name': 'SSN', 'value': '127-72-5172'}],
 '4c704ac2e3bafcd0': ['female',
  {'title': 'Ms', 'first': 'Esma', 'last': 'Ak

Отлично, у нас получился словать в котором есть все данные для таблицы

теперь создадим лист с названиями столбцов , а после и сам дата фрейм

In [39]:
columns_names = ['gender', 'name', 'location', 'email', 'login', 'dob', 'registred', 'phone', 'cell', 'id']

In [40]:
import pandas as pd

In [41]:
data = pd.DataFrame(data = dict_data, index=columns_names)
data

39c74702ac1bc239  \
gender                                                  male   
name       {'title': 'Mr', 'first': 'Sean', 'last': 'Weav...   
location   {'street': {'number': 9198, 'name': 'Shady Ln ...   
email                                sean.weaver@example.com   
login      {'uuid': 'dfaeb64c-3da9-4034-81ad-c79b109fea54...   
dob          {'date': '1972-06-19T14:17:59.978Z', 'age': 50}   
registred     {'date': '2014-08-03T07:25:36.795Z', 'age': 8}   
phone                                         (206) 996-3299   
cell                                          (851) 950-2650   
id                   {'name': 'SSN', 'value': '127-72-5172'}   

                                            4c704ac2e3bafcd0  \
gender                                                female   
name       {'title': 'Ms', 'first': 'Esma', 'last': 'Akay...   
location   {'street': {'number': 9917, 'name': 'Tunalı Hi...   
email                               esma.akaydin@example.com   
login      {'uuid': '50c0fa55-861e-4dbd-9eed-a47ec794ffe3...   
dob          {'date': '1972-05-28T10:16:39.122Z', 'age': 50}   
registred     {'date': '2021-01-11T08:25:36.127Z', 'age': 1}   
phone                                         (462)-999-1358   
cell                                          (984)-193-4509   
id                               {'name': '', 'value': None}   

                                            bb4fb271a935393b  \
gender                                                  male   
name       {'title': 'Mr', 'first': 'دانیال', 'last': 'حس...   
location   {'street': {'number': 1325, 'name': 'میدان ولی...   
email                                 dnyl.hsyny@example.com   
login      {'uuid': '51d2ceed-89a0-4907-8d8f-0fa5df273be8...   
dob          {'date': '2000-03-10T03:42:21.911Z', 'age': 22}   
registred     {'date': '2014-11-24T22:54:08.830Z', 'age': 7}   
phone                                           089-14279529   
cell                                           0932-870-3050   
id                               {'name': '', 'value': None}   

                                            f1aa6e19de4c5f35  \
gender                                                  male   
name       {'title': 'Mr', 'first': 'Lopo', 'last': 'Peix...   
location   {'street': {'number': 6382, 'name': 'Rua Paran...   
email                               lopo.peixoto@example.com   
login      {'uuid': '9b7a7d64-d990-416b-b567-58b29a4bed0c...   
dob          {'date': '1976-02-13T00:55:47.755Z', 'age': 46}   
registred     {'date': '2017-04-01T15:56:43.603Z', 'age': 5}   
phone                                         (56) 2798-0773   
cell                                          (31) 6759-5893   
id                {'name': 'CPF', 'value': '884.918.952-93'}   

                                            f62795d3059338ca  \
gender                                                  male   
name       {'title': 'Mr', 'first': 'Tilmann', 'last': 'S...   
location   {'street': {'number': 4855, 'name': 'Kapellenw...   
email                             tilmann.seelig@example.com   
login      {'uuid': 'edd77fdb-9795-4274-9e95-408a423408b1...   
dob          {'date': '1946-07-04T23:43:06.921Z', 'age': 76}   
registred    {'date': '2012-08-29T09:14:52.196Z', 'age': 10}   
phone                                           0212-3057084   
cell                                            0172-7875862   
id              {'name': 'SVNR', 'value': '19 040746 S 114'}   

                                            a72f8ae890ef4b40  \
gender                                                  male   
name       {'title': 'Mr', 'first': 'Davut', 'last': 'Akı...   
location   {'street': {'number': 8165, 'name': 'Tunalı Hi...   
email                               davut.akisik@example.com   
login      {'uuid': '410daeba-1b70-4f60-a2a5-543c7f1952ec...   
dob          {'date': '1959-04-05T10:15:06.778Z', 'age': 63}   
registred    {'date': '2007-03-28T06:19:44.763Z', 'age': 15}   
phone  

ой, наш датафрейм получился транспонированным, но ничего страшного в этом нет, есть метод transpose(), с помощью которого приведём к требуемому виду

In [42]:
data_tr =data.transpose()
data_tr

gender                                               name  \
39c74702ac1bc239    male  {'title': 'Mr', 'first': 'Sean', 'last': 'Weav...   
4c704ac2e3bafcd0  female  {'title': 'Ms', 'first': 'Esma', 'last': 'Akay...   
bb4fb271a935393b    male  {'title': 'Mr', 'first': 'دانیال', 'last': 'حس...   
f1aa6e19de4c5f35    male  {'title': 'Mr', 'first': 'Lopo', 'last': 'Peix...   
f62795d3059338ca    male  {'title': 'Mr', 'first': 'Tilmann', 'last': 'S...   
a72f8ae890ef4b40    male  {'title': 'Mr', 'first': 'Davut', 'last': 'Akı...   
7fd112b2157130ed  female   {'title': 'Mrs', 'first': 'Isla', 'last': 'Aho'}   
cc871caae4b8b48f  female  {'title': 'Mrs', 'first': 'Sonali', 'last': 'M...   
58dcbce92a0a001a  female  {'title': 'Mrs', 'first': 'Zaida', 'last': 'Di...   
b1942240f0ca789a    male  {'title': 'Mr', 'first': 'Tarak', 'last': 'Kini'}   
3319a0fc434de151  female  {'title': 'Miss', 'first': 'Shiva', 'last': 'W...   
93c3d966089ee076    male  {'title': 'Mr', 'first': 'Amador', 'last': 'Ho...   
1a04f96f90069e18  female  {'title': 'Ms', 'first': 'Siiri', 'last': 'Toro'}   
8a7a8ed9a429df61    male  {'title': 'Mr', 'first': 'محمدمهدی', 'last': '...   
85675308636fbe7e    male  {'title': 'Mr', 'first': 'Antoine', 'last': 'G...   
e232bf0460d651e3    male  {'title': 'Mr', 'first': 'سام', 'last': 'سهيلي...   
937867e6ea5d9edb    male  {'title': 'Monsieur', 'first': 'Samuel', 'last...   
349e4916c919edbb  female  {'title': 'Ms', 'first': 'Eloisa', 'last': 'Bl...   
c230b20320c3f59f    male  {'title': 'Mr', 'first': 'Riley', 'last': 'Pat...   
fed1ecb439e22534    male  {'title': 'Mr', 'first': 'Reginald', 'last': '...   

                                                           location  \
39c74702ac1bc239  {'street': {'number': 9198, 'name': 'Shady Ln ...   
4c704ac2e3bafcd0  {'street': {'number': 9917, 'name': 'Tunalı Hi...   
bb4fb271a935393b  {'street': {'number': 1325, 'name': 'میدان ولی...   
f1aa6e19de4c5f35  {'street': {'number': 6382, 'name': 'Rua Paran...   
f62795d3059338ca  {'street': {'number': 4855, 'name': 'Kapellenw...   
a72f8ae890ef4b40  {'street': {'number': 8165, 'name': 'Tunalı Hi...   
7fd112b2157130ed  {'street': {'number': 3467, 'name': 'Esplanadi...   
cc871caae4b8b48f  {'street': {'number': 8628, 'name': 'Naiduthot...   
58dcbce92a0a001a  {'street': {'number': 690, 'name': 'Rua Um'}, ...   
b1942240f0ca789a  {'street': {'number': 4009, 'name': 'Bannergha...   
3319a0fc434de151  {'street': {'number': 1299, 'name': 'Het Kromk...   
93c3d966089ee076  {'street': {'number': 8822, 'name': 'Callejón ...   
1a04f96f90069e18  {'street': {'number': 5372, 'name': 'Visiokatu...   
8a7a8ed9a429df61  {'street': {'number': 430, 'name': 'کلاهدوز'},...   
85675308636fbe7e  {'street': {'number': 9253, 'name': '20th Ave'...   
e232bf0460d651e3  {'street': {'number': 3523, 'name': 'سهروردی'}...   
937867e6ea5d9edb  {'street': {'number': 2857, 'name': 'Rue de Bo...   
349e4916c919edbb  {'street': {'number': 4102, 'name': 'Continuac...   
c230b20320c3f59f  {'street': {'number': 1009, 'name': 'Symonds S...   
fed1ecb439e22534  {'street': {'number': 7297, 'name': 'Park Aven...   

                                         email  \
39c74702ac1bc239       sean.weaver@example.com   
4c704ac2e3bafcd0      esma.akaydin@example.com   
bb4fb271a935393b        dnyl.hsyny@example.com   
f1aa6e19de4c5f35      lopo.peixoto@example.com   
f62795d3059338ca    tilmann.seelig@example.com   
a72f8ae890ef4b40      davut.akisik@example.com   
7fd112b2157130ed          isla.aho@example.com   
cc871caae4b8b48f  sonali.manjunath@example.com   
58dcbce92a0a001a        zaida.dias@example.com   
b1942240f0ca789a        tarak.kini@example.com   
3319a0fc434de151  shiva.wildenberg@example.com   
93c3d966089ee076    amador.holguin@example.com   
1a04f96f90069e18        siiri.toro@example.com   
8a7a8ed9a429df61     mhmdmhdy.rdyy@example.com   
85675308636fbe7e  antoine.gauthier@example.com   
e232bf0460d651e3        sm.shylyrd@example.com   
937867e6ea5d9edb

Избавимся от индексов

In [43]:
data = data_tr.reset_index(drop=True)
data.head(5)

gender                                               name  \
0    male  {'title': 'Mr', 'first': 'Sean', 'last': 'Weav...   
1  female  {'title': 'Ms', 'first': 'Esma', 'last': 'Akay...   
2    male  {'title': 'Mr', 'first': 'دانیال', 'last': 'حس...   
3    male  {'title': 'Mr', 'first': 'Lopo', 'last': 'Peix...   
4    male  {'title': 'Mr', 'first': 'Tilmann', 'last': 'S...   

                                            location  \
0  {'street': {'number': 9198, 'name': 'Shady Ln ...   
1  {'street': {'number': 9917, 'name': 'Tunalı Hi...   
2  {'street': {'number': 1325, 'name': 'میدان ولی...   
3  {'street': {'number': 6382, 'name': 'Rua Paran...   
4  {'street': {'number': 4855, 'name': 'Kapellenw...   

                        email  \
0     sean.weaver@example.com   
1    esma.akaydin@example.com   
2      dnyl.hsyny@example.com   
3    lopo.peixoto@example.com   
4  tilmann.seelig@example.com   

                                               login  \
0  {'uuid': 'dfaeb64c-3da9-4034-81ad-c79b109fea54...   
1  {'uuid': '50c0fa55-861e-4dbd-9eed-a47ec794ffe3...   
2  {'uuid': '51d2ceed-89a0-4907-8d8f-0fa5df273be8...   
3  {'uuid': '9b7a7d64-d990-416b-b567-58b29a4bed0c...   
4  {'uuid': 'edd77fdb-9795-4274-9e95-408a423408b1...   

                                               dob  \
0  {'date': '1972-06-19T14:17:59.978Z', 'age': 50}   
1  {'date': '1972-05-28T10:16:39.122Z', 'age': 50}   
2  {'date': '2000-03-10T03:42:21.911Z', 'age': 22}   
3  {'date': '1976-02-13T00:55:47.755Z', 'age': 46}   
4  {'date': '1946-07-04T23:43:06.921Z', 'age': 76}   

                                         registred           phone  \
0   {'date': '2014-08-03T07:25:36.795Z', 'age': 8}  (206) 996-3299   
1   {'date': '2021-01-11T08:25:36.127Z', 'age': 1}  (462)-999-1358   
2   {'date': '2014-11-24T22:54:08.830Z', 'age': 7}    089-14279529   
3   {'date': '2017-04-01T15:56:43.603Z', 'age': 5}  (56) 2798-0773   
4  {'date': '2012-08-29T09:14:52.196Z', 'age': 10}    0212-3057084   

             cell                                            id  
0  (851) 950-2650       {'name': 'SSN', 'value': '127-72-5172'}  
1  (984)-193-4509                   {'name': '', 'value': None}  
2   0932-870-3050                   {'name': '', 'value': None}  
3  (31) 6759-5893    {'name': 'CPF', 'value': '884.918.952-93'}  
4    0172-7875862  {'name': 'SVNR', 'value': '19 040746 S 114'}

### Q7

Теперь, используя библиотеку altair, сделайте два графика как на картинке:

- цвет соответствует гендеру пользователя
- первый график показывает соотношение f-m в штуках, второй - в процентах

![](http://i.imgur.com/8vyGnXR.png)

In [44]:
import altair as alt

In [45]:
num_fem = data[data['gender'] == 'female']['gender'].count()
num_fem

7

In [46]:
num_male = data[data['gender'] == 'male']['gender'].count()
num_male

13

In [47]:
total = data['gender'].count()

In [48]:
data_cnt_gender = data.groupby('gender').count().reset_index()[['gender', 'name']]
data_cnt_gender.set_axis(['gender','Count of Records'], axis = 'columns', inplace = True)
data_cnt_gender

gender  Count of Records
0  female                 7
1    male                13

In [49]:
alt.Chart(data_cnt_gender).mark_bar().encode(
    # Mapping the Website column to x-axis
    x='gender',
    # Mapping the Score column to y-axis
    y='Count of Records',
    color = 'gender'
)

alt.Chart(...)

Также без группировки можно получить такой график, используя свой датасет *data*

In [50]:
alt.Chart(data).mark_bar().encode(
    alt.X('gender'),
    alt.Y('count()'),
    alt.Color('gender')
)

alt.Chart(...)

И также процентное распределение

In [51]:
alt.Chart(data).mark_bar().encode(
    alt.X('gender', bin=True),
    alt.Y('count()', stack='normalize'),
    alt.Color('gender')
)

alt.Chart(...)

### Q8 

Найдите самого старого пользователя.

* Как его/её зовут?
* Посчитайте его/её точный возраст на **момент регистрации** в **количестве дней**

In [52]:
list_age = []
for i in range(len(data)):
    j = data['dob'][i]['age']
    list_age.append(j)

In [53]:
max(list_age)

76

Самый старый человек в нашей группе возрастом 76, найдём его и выпишем все его данные

In [54]:
k=0
for i in range(len(data)):
    j=data['dob'][i]['age']
    if j>k:
        k = j
        res = data['dob'][i]
res

{'date': '1946-07-04T23:43:06.921Z', 'age': 76}

In [55]:
old_man = data[data['dob'] == res]
old_man

gender                                               name  \
4   male  {'title': 'Mr', 'first': 'Tilmann', 'last': 'S...   

                                            location  \
4  {'street': {'number': 4855, 'name': 'Kapellenw...   

                        email  \
4  tilmann.seelig@example.com   

                                               login  \
4  {'uuid': 'edd77fdb-9795-4274-9e95-408a423408b1...   

                                               dob  \
4  {'date': '1946-07-04T23:43:06.921Z', 'age': 76}   

                                         registred         phone  \
4  {'date': '2012-08-29T09:14:52.196Z', 'age': 10}  0212-3057084   

           cell                                            id  
4  0172-7875862  {'name': 'SVNR', 'value': '19 040746 S 114'}

In [56]:
name = old_man['name'][4]['title'] +' '+ old_man['name'][4]['first'] +' '+ old_man['name'][4]['last']
name

'Mr Tilmann Seelig'

In [57]:
import datetime

Вытащим даты между которыми будем искать разницу времени

Но сразу можно заметить, что формат таких записей будут строки, значит будем разделять и переводить в формат времени datetime

In [58]:
date_regist = old_man['registred'][4]['date']
date_regist_obj = datetime.datetime.strptime(date_regist, '%Y-%m-%dT%H:%M:%S.%fZ')
date_regist_obj

datetime.datetime(2012, 8, 29, 9, 14, 52, 196000)

In [59]:
date_dob = old_man['dob'][4]['date']
date_dob_obj = datetime.datetime.strptime(date_dob, '%Y-%m-%dT%H:%M:%S.%fZ')
date_dob_obj

datetime.datetime(1946, 7, 4, 23, 43, 6, 921000)

Разница в днях будет вычисляться как:

In [60]:
date_regist_obj- date_dob_obj

datetime.timedelta(days=24162, seconds=34305, microseconds=275000)

In [61]:
days_diff = (date_regist_obj - date_dob_obj).days
days_diff

24162

24162 дня было *'Mr Tilmann Seelig'* на момент регистрации, совсем юнец :)

### Q9

используя `pd.DataFrame.groupby`, посчитайте:
* для мужчин VS женщин
    * их количество
    * средний возраст
    * стандартное отклонениe возраста
    * среднее кол-во лет с момента регистрации
    * стандартное отклонениe кол-вa лет с момента регистрации

In [62]:
data.head(5)

gender                                               name  \
0    male  {'title': 'Mr', 'first': 'Sean', 'last': 'Weav...   
1  female  {'title': 'Ms', 'first': 'Esma', 'last': 'Akay...   
2    male  {'title': 'Mr', 'first': 'دانیال', 'last': 'حس...   
3    male  {'title': 'Mr', 'first': 'Lopo', 'last': 'Peix...   
4    male  {'title': 'Mr', 'first': 'Tilmann', 'last': 'S...   

                                            location  \
0  {'street': {'number': 9198, 'name': 'Shady Ln ...   
1  {'street': {'number': 9917, 'name': 'Tunalı Hi...   
2  {'street': {'number': 1325, 'name': 'میدان ولی...   
3  {'street': {'number': 6382, 'name': 'Rua Paran...   
4  {'street': {'number': 4855, 'name': 'Kapellenw...   

                        email  \
0     sean.weaver@example.com   
1    esma.akaydin@example.com   
2      dnyl.hsyny@example.com   
3    lopo.peixoto@example.com   
4  tilmann.seelig@example.com   

                                               login  \
0  {'uuid': 'dfaeb64c-3da9-4034-81ad-c79b109fea54...   
1  {'uuid': '50c0fa55-861e-4dbd-9eed-a47ec794ffe3...   
2  {'uuid': '51d2ceed-89a0-4907-8d8f-0fa5df273be8...   
3  {'uuid': '9b7a7d64-d990-416b-b567-58b29a4bed0c...   
4  {'uuid': 'edd77fdb-9795-4274-9e95-408a423408b1...   

                                               dob  \
0  {'date': '1972-06-19T14:17:59.978Z', 'age': 50}   
1  {'date': '1972-05-28T10:16:39.122Z', 'age': 50}   
2  {'date': '2000-03-10T03:42:21.911Z', 'age': 22}   
3  {'date': '1976-02-13T00:55:47.755Z', 'age': 46}   
4  {'date': '1946-07-04T23:43:06.921Z', 'age': 76}   

                                         registred           phone  \
0   {'date': '2014-08-03T07:25:36.795Z', 'age': 8}  (206) 996-3299   
1   {'date': '2021-01-11T08:25:36.127Z', 'age': 1}  (462)-999-1358   
2   {'date': '2014-11-24T22:54:08.830Z', 'age': 7}    089-14279529   
3   {'date': '2017-04-01T15:56:43.603Z', 'age': 5}  (56) 2798-0773   
4  {'date': '2012-08-29T09:14:52.196Z', 'age': 10}    0212-3057084   

             cell                                            id  
0  (851) 950-2650       {'name': 'SSN', 'value': '127-72-5172'}  
1  (984)-193-4509                   {'name': '', 'value': None}  
2   0932-870-3050                   {'name': '', 'value': None}  
3  (31) 6759-5893    {'name': 'CPF', 'value': '884.918.952-93'}  
4    0172-7875862  {'name': 'SVNR', 'value': '19 040746 S 114'}

Для удобной работы с таблицей выведем в новые столбцы возраст и даты регистрации с датами рождения

In [63]:
data_new = data.copy()

In [64]:
data_new['age'] = data_new['dob']

In [65]:
for i in range(len(data_new)):
    data_new['age'][i] = data_new['dob'][i]['age']
    data_new['dob'][i] = data_new['dob'][i]['date']

In [66]:
#data2['first_day_exposition'] = pd.to_datetime(data['first_day_exposition'], format='%Y-%m-%dT%H:%M:%S')
data_new.head(5)

gender                                               name  \
0    male  {'title': 'Mr', 'first': 'Sean', 'last': 'Weav...   
1  female  {'title': 'Ms', 'first': 'Esma', 'last': 'Akay...   
2    male  {'title': 'Mr', 'first': 'دانیال', 'last': 'حس...   
3    male  {'title': 'Mr', 'first': 'Lopo', 'last': 'Peix...   
4    male  {'title': 'Mr', 'first': 'Tilmann', 'last': 'S...   

                                            location  \
0  {'street': {'number': 9198, 'name': 'Shady Ln ...   
1  {'street': {'number': 9917, 'name': 'Tunalı Hi...   
2  {'street': {'number': 1325, 'name': 'میدان ولی...   
3  {'street': {'number': 6382, 'name': 'Rua Paran...   
4  {'street': {'number': 4855, 'name': 'Kapellenw...   

                        email  \
0     sean.weaver@example.com   
1    esma.akaydin@example.com   
2      dnyl.hsyny@example.com   
3    lopo.peixoto@example.com   
4  tilmann.seelig@example.com   

                                               login  \
0  {'uuid': 'dfaeb64c-3da9-4034-81ad-c79b109fea54...   
1  {'uuid': '50c0fa55-861e-4dbd-9eed-a47ec794ffe3...   
2  {'uuid': '51d2ceed-89a0-4907-8d8f-0fa5df273be8...   
3  {'uuid': '9b7a7d64-d990-416b-b567-58b29a4bed0c...   
4  {'uuid': 'edd77fdb-9795-4274-9e95-408a423408b1...   

                        dob                                        registred  \
0  1972-06-19T14:17:59.978Z   {'date': '2014-08-03T07:25:36.795Z', 'age': 8}   
1  1972-05-28T10:16:39.122Z   {'date': '2021-01-11T08:25:36.127Z', 'age': 1}   
2  2000-03-10T03:42:21.911Z   {'date': '2014-11-24T22:54:08.830Z', 'age': 7}   
3  1976-02-13T00:55:47.755Z   {'date': '2017-04-01T15:56:43.603Z', 'age': 5}   
4  1946-07-04T23:43:06.921Z  {'date': '2012-08-29T09:14:52.196Z', 'age': 10}   

            phone            cell  \
0  (206) 996-3299  (851) 950-2650   
1  (462)-999-1358  (984)-193-4509   
2    089-14279529   0932-870-3050   
3  (56) 2798-0773  (31) 6759-5893   
4    0212-3057084    0172-7875862   

                                             id age  
0       {'name': 'SSN', 'value': '127-72-5172'}  50  
1                   {'name': '', 'value': None}  50  
2                   {'name': '', 'value': None}  22  
3    {'name': 'CPF', 'value': '884.918.952-93'}  46  
4  {'name': 'SVNR', 'value': '19 040746 S 114'}  76

In [67]:
data_new['dob'] = pd.to_datetime(data_new['dob'], format='%Y-%m-%dT%H:%M:%S.%fZ')

In [68]:
data_new['data_registred'] = data_new['registred']
data_new

gender                                               name  \
0     male  {'title': 'Mr', 'first': 'Sean', 'last': 'Weav...   
1   female  {'title': 'Ms', 'first': 'Esma', 'last': 'Akay...   
2     male  {'title': 'Mr', 'first': 'دانیال', 'last': 'حس...   
3     male  {'title': 'Mr', 'first': 'Lopo', 'last': 'Peix...   
4     male  {'title': 'Mr', 'first': 'Tilmann', 'last': 'S...   
5     male  {'title': 'Mr', 'first': 'Davut', 'last': 'Akı...   
6   female   {'title': 'Mrs', 'first': 'Isla', 'last': 'Aho'}   
7   female  {'title': 'Mrs', 'first': 'Sonali', 'last': 'M...   
8   female  {'title': 'Mrs', 'first': 'Zaida', 'last': 'Di...   
9     male  {'title': 'Mr', 'first': 'Tarak', 'last': 'Kini'}   
10  female  {'title': 'Miss', 'first': 'Shiva', 'last': 'W...   
11    male  {'title': 'Mr', 'first': 'Amador', 'last': 'Ho...   
12  female  {'title': 'Ms', 'first': 'Siiri', 'last': 'Toro'}   
13    male  {'title': 'Mr', 'first': 'محمدمهدی', 'last': '...   
14    male  {'title': 'Mr', 'first': 'Antoine', 'last': 'G...   
15    male  {'title': 'Mr', 'first': 'سام', 'last': 'سهيلي...   
16    male  {'title': 'Monsieur', 'first': 'Samuel', 'last...   
17  female  {'title': 'Ms', 'first': 'Eloisa', 'last': 'Bl...   
18    male  {'title': 'Mr', 'first': 'Riley', 'last': 'Pat...   
19    male  {'title': 'Mr', 'first': 'Reginald', 'last': '...   

                                             location  \
0   {'street': {'number': 9198, 'name': 'Shady Ln ...   
1   {'street': {'number': 9917, 'name': 'Tunalı Hi...   
2   {'street': {'number': 1325, 'name': 'میدان ولی...   
3   {'street': {'number': 6382, 'name': 'Rua Paran...   
4   {'street': {'number': 4855, 'name': 'Kapellenw...   
5   {'street': {'number': 8165, 'name': 'Tunalı Hi...   
6   {'street': {'number': 3467, 'name': 'Esplanadi...   
7   {'street': {'number': 8628, 'name': 'Naiduthot...   
8   {'street': {'number': 690, 'name': 'Rua Um'}, ...   
9   {'street': {'number': 4009, 'name': 'Bannergha...   
10  {'street': {'number': 1299, 'name': 'Het Kromk...   
11  {'street': {'number': 8822, 'name': 'Callejón ...   
12  {'street': {'number': 5372, 'name': 'Visiokatu...   
13  {'street': {'number': 430, 'name': 'کلاهدوز'},...   
14  {'street': {'number': 9253, 'name': '20th Ave'...   
15  {'street': {'number': 3523, 'name': 'سهروردی'}...   
16  {'street': {'number': 2857, 'name': 'Rue de Bo...   
17  {'street': {'number': 4102, 'name': 'Continuac...   
18  {'street': {'number': 1009, 'name': 'Symonds S...   
19  {'street': {'number': 7297, 'name': 'Park Aven...   

                           email  \
0        sean.weaver@example.com   
1       esma.akaydin@example.com   
2         dnyl.hsyny@example.com   
3       lopo.peixoto@example.com   
4     tilmann.seelig@example.com   
5       davut.akisik@example.com   
6           isla.aho@example.com   
7   sonali.manjunath@example.com   
8         zaida.dias@example.com   
9         tarak.kini@example.com   
10  shiva.wildenberg@example.com   
11    amador.holguin@example.com   
12        siiri.toro@example.com   
13     mhmdmhdy.rdyy@example.com   
14  antoine.gauthier@example.com   
15        sm.shylyrd@example.com   
16       samuel.adam@example.com   
17     eloisa.blanco@example.com   
18       riley.patel@example.com   
19    reginald.mason@example.com   

                                                login                     dob  \
0   {'uuid': 'dfaeb64c-3da9-4034-81ad-c79b109fea54... 1972-06-19 14:17:59.978   
1   {'uuid': '50c0fa55-861e-4dbd-9eed-a47ec794ffe3... 1972-05-28 10:16:39.122   
2   {'uuid': '51d2ceed-89a0-4907-8d8f-0fa5df273be8... 2000-03-10 03:42:21.911   
3   {'uuid': '9b7a7d64-d990-416b-b567-58b29a4bed0c... 1976-02-13 00:55:47.755   
4   {'uuid': 'edd77fdb-9795-4274-9e95-408a423408b1... 1946-07-04 23:43:06.921   
5   {'uuid': '410daeba-1b70-4f60-a2a5-543c7f1952ec... 1959-04-05 10:15:06.778   
6   {'uuid': '605551d0-e9b0-47e7-b317-1bbda214ca19... 1984-08-29 14:30:03.172   
7   {'uuid': 'd78bd98d-8692-492b-b5

In [69]:
list_date_ = []
for i in data_new['registred']:
    list_date_.append(i['date'])

In [70]:
series_date = pd.Series(list_date_)

In [71]:
data_new['data_registred'] = series_date
data_new['data_registred']

0     2014-08-03T07:25:36.795Z
1     2021-01-11T08:25:36.127Z
2     2014-11-24T22:54:08.830Z
3     2017-04-01T15:56:43.603Z
4     2012-08-29T09:14:52.196Z
5     2007-03-28T06:19:44.763Z
6     2019-05-05T06:45:03.155Z
7     2022-01-15T00:41:58.570Z
8     2018-11-24T10:39:08.982Z
9     2004-04-30T06:54:52.830Z
10    2015-05-11T17:49:44.315Z
11    2015-10-13T02:54:24.554Z
12    2003-09-26T00:06:56.319Z
13    2010-09-01T15:26:32.668Z
14    2012-04-23T02:18:21.092Z
15    2016-12-06T04:34:19.118Z
16    2021-11-17T05:24:46.536Z
17    2021-08-18T22:25:35.082Z
18    2020-06-03T01:13:29.386Z
19    2005-04-09T15:07:39.474Z
Name: data_registred, dtype: object

In [72]:
data_new['data_registred'] = pd.to_datetime(data_new['data_registred'], format='%Y-%m-%dT%H:%M:%S.%fZ')

In [73]:
data_new.head(5)

gender                                               name  \
0    male  {'title': 'Mr', 'first': 'Sean', 'last': 'Weav...   
1  female  {'title': 'Ms', 'first': 'Esma', 'last': 'Akay...   
2    male  {'title': 'Mr', 'first': 'دانیال', 'last': 'حس...   
3    male  {'title': 'Mr', 'first': 'Lopo', 'last': 'Peix...   
4    male  {'title': 'Mr', 'first': 'Tilmann', 'last': 'S...   

                                            location  \
0  {'street': {'number': 9198, 'name': 'Shady Ln ...   
1  {'street': {'number': 9917, 'name': 'Tunalı Hi...   
2  {'street': {'number': 1325, 'name': 'میدان ولی...   
3  {'street': {'number': 6382, 'name': 'Rua Paran...   
4  {'street': {'number': 4855, 'name': 'Kapellenw...   

                        email  \
0     sean.weaver@example.com   
1    esma.akaydin@example.com   
2      dnyl.hsyny@example.com   
3    lopo.peixoto@example.com   
4  tilmann.seelig@example.com   

                                               login                     dob  \
0  {'uuid': 'dfaeb64c-3da9-4034-81ad-c79b109fea54... 1972-06-19 14:17:59.978   
1  {'uuid': '50c0fa55-861e-4dbd-9eed-a47ec794ffe3... 1972-05-28 10:16:39.122   
2  {'uuid': '51d2ceed-89a0-4907-8d8f-0fa5df273be8... 2000-03-10 03:42:21.911   
3  {'uuid': '9b7a7d64-d990-416b-b567-58b29a4bed0c... 1976-02-13 00:55:47.755   
4  {'uuid': 'edd77fdb-9795-4274-9e95-408a423408b1... 1946-07-04 23:43:06.921   

                                         registred           phone  \
0   {'date': '2014-08-03T07:25:36.795Z', 'age': 8}  (206) 996-3299   
1   {'date': '2021-01-11T08:25:36.127Z', 'age': 1}  (462)-999-1358   
2   {'date': '2014-11-24T22:54:08.830Z', 'age': 7}    089-14279529   
3   {'date': '2017-04-01T15:56:43.603Z', 'age': 5}  (56) 2798-0773   
4  {'date': '2012-08-29T09:14:52.196Z', 'age': 10}    0212-3057084   

             cell                                            id age  \
0  (851) 950-2650       {'name': 'SSN', 'value': '127-72-5172'}  50   
1  (984)-193-4509                   {'name': '', 'value': None}  50   
2   0932-870-3050                   {'name': '', 'value': None}  22   
3  (31) 6759-5893    {'name': 'CPF', 'value': '884.918.952-93'}  46   
4    0172-7875862  {'name': 'SVNR', 'value': '19 040746 S 114'}  76   

           data_registred  
0 2014-08-03 07:25:36.795  
1 2021-01-11 08:25:36.127  
2 2014-11-24 22:54:08.830  
3 2017-04-01 15:56:43.603  
4 2012-08-29 09:14:52.196

In [74]:
data_new_2 = data_new[['gender','dob','age','data_registred']]
data_new_2.head(5)

gender                     dob age          data_registred
0    male 1972-06-19 14:17:59.978  50 2014-08-03 07:25:36.795
1  female 1972-05-28 10:16:39.122  50 2021-01-11 08:25:36.127
2    male 2000-03-10 03:42:21.911  22 2014-11-24 22:54:08.830
3    male 1976-02-13 00:55:47.755  46 2017-04-01 15:56:43.603
4    male 1946-07-04 23:43:06.921  76 2012-08-29 09:14:52.196

Отлично, теперь есть с чем работать

**Количество мужчин и женщин**

In [75]:
data_group = data_new_2.groupby('gender')['dob'].count()
data_group

gender
female     7
male      13
Name: dob, dtype: int64

**Средний возраст по категориям**

In [76]:
data_group = data_new_2.groupby('gender')['age'].mean()
data_group

gender
female    51.142857
male      47.000000
Name: age, dtype: float64

**Стандартное отклонение возраста**

In [77]:
data_group = data_new_2.groupby('gender')['age'].std()
data_group

gender
female    10.869354
male      19.437935
Name: age, dtype: float64

In [78]:
import datetime

In [79]:
data_new_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   gender          20 non-null     object        
 1   dob             20 non-null     datetime64[ns]
 2   age             20 non-null     object        
 3   data_registred  20 non-null     datetime64[ns]
dtypes: datetime64[ns](2), object(2)
memory usage: 768.0+ bytes


In [80]:
data_new_2['diff_age'] = data_new_2.apply(lambda x: round((x['data_registred']-x['dob']).days/365,2), axis = 1)

C:\Users\ruk19\AppData\Local\Temp/ipykernel_13992/1531908096.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_new_2['diff_age'] = data_new_2.apply(lambda x: round((x['data_registred']-x['dob']).days/365,2), axis = 1)


In [81]:
data_new_2

gender                     dob age          data_registred  diff_age
0     male 1972-06-19 14:17:59.978  50 2014-08-03 07:25:36.795     42.15
1   female 1972-05-28 10:16:39.122  50 2021-01-11 08:25:36.127     48.65
2     male 2000-03-10 03:42:21.911  22 2014-11-24 22:54:08.830     14.72
3     male 1976-02-13 00:55:47.755  46 2017-04-01 15:56:43.603     41.16
4     male 1946-07-04 23:43:06.921  76 2012-08-29 09:14:52.196     66.20
5     male 1959-04-05 10:15:06.778  63 2007-03-28 06:19:44.763     48.01
6   female 1984-08-29 14:30:03.172  38 2019-05-05 06:45:03.155     34.70
7   female 1974-06-14 19:26:19.276  48 2022-01-15 00:41:58.570     47.62
8   female 1966-10-08 12:16:18.429  56 2018-11-24 10:39:08.982     52.16
9     male 1956-06-22 05:14:00.152  66 2004-04-30 06:54:52.830     47.89
10  female 1979-05-24 02:23:44.420  43 2015-05-11 17:49:44.315     35.99
11    male 1986-05-11 07:09:57.335  36 2015-10-13 02:54:24.554     29.44
12  female 1950-06-05 16:18:33.872  72 2003-09-26 00:06:56.319     53.34
13    male 1987-11-05 02:27:18.022  34 2010-09-01 15:26:32.668     22.84
14    male 1956-07-03 18:26:52.065  66 2012-04-23 02:18:21.092     55.84
15    male 1993-05-24 09:18:35.480  29 2016-12-06 04:34:19.118     23.55
16    male 1998-04-09 09:28:38.907  24 2021-11-17 05:24:46.536     23.62
17  female 1971-10-10 22:02:16.392  51 2021-08-18 22:25:35.082     49.89
18    male 1994-09-26 14:06:37.957  28 2020-06-03 01:13:29.386     25.70
19    male 1951-08-06 17:20:30.109  71 2005-04-09 15:07:39.474     53.71

**Среднее количество лет с момента регистрации**

In [82]:
data_group = data_new_2.groupby('gender')['diff_age'].mean()
data_group

gender
female    46.050000
male      38.063846
Name: diff_age, dtype: float64

**Стандартное отклонениe кол-вa лет с момента регистрации**

In [83]:
data_group = data_new_2.groupby('gender')['diff_age'].std()
data_group

gender
female     7.577079
male      15.804566
Name: diff_age, dtype: float64

### Q10

 Удалите всех пользователей, зарегестрированных раньше 2012 или после 2018
 
 Выражаясь математически, оставьте только промежуток `[2012, 2018)`

In [84]:
data_new_2['years_reg'] = data_new_2.apply(lambda x: x['data_registred'].year, axis=1)

C:\Users\ruk19\AppData\Local\Temp/ipykernel_13992/2890525193.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_new_2['years_reg'] = data_new_2.apply(lambda x: x['data_registred'].year, axis=1)


In [85]:
data_new_2['years_reg']

0     2014
1     2021
2     2014
3     2017
4     2012
5     2007
6     2019
7     2022
8     2018
9     2004
10    2015
11    2015
12    2003
13    2010
14    2012
15    2016
16    2021
17    2021
18    2020
19    2005
Name: years_reg, dtype: int64

In [86]:
data_new_3 = data_new_2.query('years_reg<2012 or years_reg>2018')
data_new_3

gender                     dob age          data_registred  diff_age  \
1   female 1972-05-28 10:16:39.122  50 2021-01-11 08:25:36.127     48.65   
5     male 1959-04-05 10:15:06.778  63 2007-03-28 06:19:44.763     48.01   
6   female 1984-08-29 14:30:03.172  38 2019-05-05 06:45:03.155     34.70   
7   female 1974-06-14 19:26:19.276  48 2022-01-15 00:41:58.570     47.62   
9     male 1956-06-22 05:14:00.152  66 2004-04-30 06:54:52.830     47.89   
12  female 1950-06-05 16:18:33.872  72 2003-09-26 00:06:56.319     53.34   
13    male 1987-11-05 02:27:18.022  34 2010-09-01 15:26:32.668     22.84   
16    male 1998-04-09 09:28:38.907  24 2021-11-17 05:24:46.536     23.62   
17  female 1971-10-10 22:02:16.392  51 2021-08-18 22:25:35.082     49.89   
18    male 1994-09-26 14:06:37.957  28 2020-06-03 01:13:29.386     25.70   
19    male 1951-08-06 17:20:30.109  71 2005-04-09 15:07:39.474     53.71   

    years_reg  
1        2021  
5        2007  
6        2019  
7        2022  
9        2004  
12       2003  
13       2010  
16       2021  
17       2021  
18       2020  
19       2005

Теперь соединим таблицы по индексам и останутся только нужные нам значения

In [87]:
data_2012_2018 = pd.concat([data, data_new_3], axis =1).dropna()
data_2012_2018

gender                                               name  \
1   female  {'title': 'Ms', 'first': 'Esma', 'last': 'Akay...   
5     male  {'title': 'Mr', 'first': 'Davut', 'last': 'Akı...   
6   female   {'title': 'Mrs', 'first': 'Isla', 'last': 'Aho'}   
7   female  {'title': 'Mrs', 'first': 'Sonali', 'last': 'M...   
9     male  {'title': 'Mr', 'first': 'Tarak', 'last': 'Kini'}   
12  female  {'title': 'Ms', 'first': 'Siiri', 'last': 'Toro'}   
13    male  {'title': 'Mr', 'first': 'محمدمهدی', 'last': '...   
16    male  {'title': 'Monsieur', 'first': 'Samuel', 'last...   
17  female  {'title': 'Ms', 'first': 'Eloisa', 'last': 'Bl...   
18    male  {'title': 'Mr', 'first': 'Riley', 'last': 'Pat...   
19    male  {'title': 'Mr', 'first': 'Reginald', 'last': '...   

                                             location  \
1   {'street': {'number': 9917, 'name': 'Tunalı Hi...   
5   {'street': {'number': 8165, 'name': 'Tunalı Hi...   
6   {'street': {'number': 3467, 'name': 'Esplanadi...   
7   {'street': {'number': 8628, 'name': 'Naiduthot...   
9   {'street': {'number': 4009, 'name': 'Bannergha...   
12  {'street': {'number': 5372, 'name': 'Visiokatu...   
13  {'street': {'number': 430, 'name': 'کلاهدوز'},...   
16  {'street': {'number': 2857, 'name': 'Rue de Bo...   
17  {'street': {'number': 4102, 'name': 'Continuac...   
18  {'street': {'number': 1009, 'name': 'Symonds S...   
19  {'street': {'number': 7297, 'name': 'Park Aven...   

                           email  \
1       esma.akaydin@example.com   
5       davut.akisik@example.com   
6           isla.aho@example.com   
7   sonali.manjunath@example.com   
9         tarak.kini@example.com   
12        siiri.toro@example.com   
13     mhmdmhdy.rdyy@example.com   
16       samuel.adam@example.com   
17     eloisa.blanco@example.com   
18       riley.patel@example.com   
19    reginald.mason@example.com   

                                                login  \
1   {'uuid': '50c0fa55-861e-4dbd-9eed-a47ec794ffe3...   
5   {'uuid': '410daeba-1b70-4f60-a2a5-543c7f1952ec...   
6   {'uuid': '605551d0-e9b0-47e7-b317-1bbda214ca19...   
7   {'uuid': 'd78bd98d-8692-492b-b560-7e8fd7b558eb...   
9   {'uuid': 'e676021c-2b00-452e-97d6-c3174857b317...   
12  {'uuid': 'eef28f95-3a47-499b-baf9-d8113689f283...   
13  {'uuid': 'c2b0f1ee-3ff8-4bad-89f4-bc6a6a6c16ae...   
16  {'uuid': 'd054716a-ab7c-49b6-8922-effa1e47fe93...   
17  {'uuid': '51dd67dc-3401-4374-977d-ae7f97c4f475...   
18  {'uuid': 'd9a1d3cb-cd7c-4ccc-912f-8eee7a1752c4...   
19  {'uuid': 'c97b9dea-1d34-4207-a6e2-812934dec2dd...   

                                                dob  \
1   {'date': '1972-05-28T10:16:39.122Z', 'age': 50}   
5   {'date': '1959-04-05T10:15:06.778Z', 'age': 63}   
6   {'date': '1984-08-29T14:30:03.172Z', 'age': 38}   
7   {'date': '1974-06-14T19:26:19.276Z', 'age': 48}   
9   {'date': '1956-06-22T05:14:00.152Z', 'age': 66}   
12  {'date': '1950-06-05T16:18:33.872Z', 'age': 72}   
13  {'date': '1987-11-05T02:27:18.022Z', 'age': 34}   
16  {'date': '1998-04-09T09:28:38.907Z', 'age': 24}   
17  {'date': '1971-10-10T22:02:16.392Z', 'age': 51}   
18  {'date': '1994-09-26T14:06:37.957Z', 'age': 28}   
19  {'date': '1951-08-06T17:20:30.109Z', 'age': 71}   

                                          registred           phone  \
1    {'date': '2021-01-11T08:25:36.127Z', 'age': 1}  (462)-999-1358   
5   {'date': '2007-03-28T06:19:44.763Z', 'age': 15}  (491)-531-9734   
6    {'date': '2019-05-05T06:45:03.155Z', 'age': 3}      03-037-616   
7    {'date': '2022-01-15T00:41:58.570Z', 'age': 0}      8633597217   
9   {'date': '2004-04-30T06:54:52.830Z', 'age': 18}      7534610427   
12  {'date': '2003-09-26T00:06:56.319Z', 'age': 19}      05-559-152   
13  {'date': '2010-09-01T15:26:32.668Z', 'age': 12}    046-47046925   
16   {'date': '2021-11-17T05:24:46.536Z', 'age': 0}   079 249 95 26   
17   {'date': '2021-08-18T22:25:35.082Z', 'age': 1}  (647) 292 4324   
18   {'date': '2020-06-03T01:13:29.386Z', 'age': 

In [88]:
data_2012_2018.shape

(11, 16)

In [89]:
data_2012_2018.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11 entries, 1 to 19
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   gender          11 non-null     object        
 1   name            11 non-null     object        
 2   location        11 non-null     object        
 3   email           11 non-null     object        
 4   login           11 non-null     object        
 5   dob             11 non-null     object        
 6   registred       11 non-null     object        
 7   phone           11 non-null     object        
 8   cell            11 non-null     object        
 9   id              11 non-null     object        
 10  gender          11 non-null     object        
 11  dob             11 non-null     datetime64[ns]
 12  age             11 non-null     object        
 13  data_registred  11 non-null     datetime64[ns]
 14  diff_age        11 non-null     float64       
 15  years_re